### PR TITLE
tikv_util: make memory quota alloc and free simpler

### PR DESCRIPTION
**Pull request by glorv**
### What is changed and how it works?

Issue Number: ref #15990

What's Changed:

```commit-message
Replace `compare_exchange_weak` loop with single `fetch_add` and `fetch_sub` to make the memory alloc and free simpler.
I do this change base on the fact that we does not expect the memory control to be precise on every operation. In high concurrency scenario and when the conflict is heavy, the real peak memory usage may be small than the capacity, but not smaller than `capacity - num_of_concurrency * avg_alloc_bytes`.

NOTE: In order to avoid counter overflow, this change introduces a hard limit(256PiB) that one alloc can acquire as it's big enough for any real world application. Passing higher value to `alloc` will result in an error, passing higher value to `alloc_force` and `free` will result in noop.
```

benchmark Result: (The benchmark was running with a modified version of  https://github.com/overvenus/benches-rust/blob/82c75f87bb7b9fabdbb2bbcf2bb11cc8f726d3a8/benches/memory.rs to compare the old and new version of MemoryQuota)
```
Benchmarking Alloc Only/Alloc/ok: Collecting 100 samples in estimated 5.0000 s --  Alloc Only/Alloc/ok     
                        time:   [11.627 ns 11.707 ns 11.783 ns]
Benchmarking Alloc Only New/AllocNew/ok: Collecting 100 samples in estimated..  -- Alloc Only New/AllocNew/ok   
                        time:   [13.263 ns 15.141 ns 17.126 ns]


Benchmarking Alloc Only/Alloc/fail: Collecting 100 samples in estimated 5.0000 s -- Alloc Only/Alloc/fail   
                        time:   [2.7350 ns 2.7457 ns 2.7566 ns]
Benchmarking Alloc Only New/AllocNew/fail: Collecting 100 samples in estimated 5 -- Alloc Only New/AllocNew/fail    
                        time:   [2.5062 ns 2.5333 ns 2.5629 ns]


Benchmarking Alloc Free/MemoryQuota/alloc free: Collecting 100 samples in esti.. --  Alloc Free/MemoryQuota/alloc free
                        time:   [22.268 ns 22.455 ns 22.656 ns]
Benchmarking Alloc Free New/MemoryQuotaNew/alloc free: Collecting 100 samples .. -- Alloc Free New/MemoryQuotaNew/alloc free
                        time:   [19.653 ns 19.878 ns 20.154 ns]


Benchmarking Alloc Free/OwnedAllocated/alloc free: Collecting 100 samples in est -- Alloc Free/OwnedAllocated/alloc free
                        time:   [34.812 ns 35.084 ns 35.397 ns]
Benchmarking Alloc Free New/OwnedAllocatedNew/alloc free: Collecting 100 samples -- Alloc Free New/OwnedAllocatedNew/alloc free
                        time:   [35.152 ns 37.407 ns 40.522 ns]


Benchmarking 2 Threads/MemoryQuota/alloc free: Collecting 100 samples in estimat -- 2 Threads/MemoryQuota/alloc free
                        time:   [133.58 ns 136.84 ns 140.07 ns]
Benchmarking 2 Threads New/MemoryQuota/alloc free: Collecting 100 samples in est -- 2 Threads New/MemoryQuota/alloc free
                        time:   [23.941 ns 24.808 ns 25.850 ns]


Benchmarking 2 Threads/OwnedAllocated/alloc free: Collecting 100 samples in esti -- 2 Threads/OwnedAllocated/alloc free
                        time:   [208.20 ns 214.79 ns 221.67 ns]
Benchmarking 2 Threads New/OwnedAllocated/alloc free: Collecting 100 samples in  -- 2 Threads New/OwnedAllocated/alloc free
                        time:   [41.930 ns 45.758 ns 51.506 ns]

Benchmarking 4 Threads/MemoryQuota/alloc free: Collecting 100 samples in estimat -- 4 Threads/MemoryQuota/alloc free
                        time:   [421.70 ns 435.66 ns 452.09 ns]
Benchmarking 4 Threads New/MemoryQuota/alloc free: Collecting 100 samples in est -- 4 Threads New/MemoryQuota/alloc free
                        time:   [38.904 ns 42.572 ns 48.319 ns]


Benchmarking 4 Threads/OwnedAllocated/alloc free: Collecting 100 samples in esti -- 4 Threads/OwnedAllocated/alloc free
                        time:   [619.40 ns 634.09 ns 651.68 ns]
Benchmarking 4 Threads New/OwnedAllocated/alloc free: Collecting 100 samples in  -- 4 Threads New/OwnedAllocated/alloc free
                        time:   [54.774 ns 56.300 ns 57.955 ns]


Benchmarking 8 Threads/MemoryQuota/alloc free: Collecting 100 samples in estimat -- 8 Threads/MemoryQuota/alloc free
                        time:   [1.9759 µs 2.0775 µs 2.1650 µs]
Benchmarking 8 Threads New/MemoryQuota/alloc free: Collecting 100 samples in est -- 8 Threads New/MemoryQuota/alloc free
                        time:   [119.88 ns 154.39 ns 199.66 ns]


Benchmarking 8 Threads/OwnedAllocated/alloc free: Collecting 100 samples in esti -- 8 Threads/OwnedAllocated/alloc free
                        time:   [2.7435 µs 2.8082 µs 2.8620 µs]
Benchmarking 8 Threads New/OwnedAllocated/alloc free: Collecting 100 samples in  -- 8 Threads New/OwnedAllocated/alloc free
                        time:   [390.64 ns 461.45 ns 539.01 ns]


Benchmarking 16 Threads/MemoryQuota/alloc free: Collecting 100 samples in estima -- 16 Threads/MemoryQuota/alloc free
                        time:   [1.0091 µs 1.0961 µs 1.1982 µs]
Benchmarking 16 Threads New/MemoryQuota/alloc free: Collecting 100 samples in es -- 16 Threads New/MemoryQuota/alloc free
                        time:   [263.66 ns 305.28 ns 353.69 ns]


Benchmarking 16 Threads/OwnedAllocated/alloc free: Collecting 100 samples in est -- 16 Threads/OwnedAllocated/alloc free
                        time:   [1.1606 µs 1.1975 µs 1.2382 µs]
Benchmarking 16 Threads New/OwnedAllocated/alloc free: Collecting 100 samples in -- 16 Threads New/OwnedAllocated/alloc free
                        time:   [166.27 ns 172.57 ns 181.63 ns]


Benchmarking 32 Threads/MemoryQuota/alloc free: Collecting 100 samples in estima -- 32 Threads/MemoryQuota/alloc free
                        time:   [1.1244 µs 1.2449 µs 1.3840 µs]
Benchmarking 32 Threads New/MemoryQuota/alloc free: Collecting 100 samples in es -- 32 Threads New/MemoryQuota/alloc free
                        time:   [428.08 ns 500.57 ns 582.01 ns]


Benchmarking 32 Threads/OwnedAllocated/alloc free: Collecting 100 samples in est -- 32 Threads/OwnedAllocated/alloc free
                        time:   [1.9411 µs 2.1136 µs 2.3348 µs]
Benchmarking 32 Threads New/OwnedAllocated/alloc free: Collecting 100 samples in -- 32 Threads New/OwnedAllocated/alloc free
                        time:   [541.50 ns 603.69 ns 661.78 ns]


Benchmarking 64 Threads/MemoryQuota/alloc free: Collecting 100 samples in estima -- 64 Threads/MemoryQuota/alloc free
                        time:   [1.9853 µs 2.2539 µs 2.5148 µs]
Benchmarking 64 Threads New/MemoryQuota/alloc free: Collecting 100 samples in es -- 64 Threads New/MemoryQuota/alloc free
                        time:   [802.77 ns 852.78 ns 895.60 ns]



Benchmarking 64 Threads/OwnedAllocated/alloc free: Collecting 100 samples in est -- 64 Threads/OwnedAllocated/alloc free
                        time:   [3.5248 µs 3.7999 µs 4.1271 µs]
Benchmarking 64 Threads New/OwnedAllocated/alloc free: Collecting 100 samples in -- 64 Threads New/OwnedAllocated/alloc free
                        time:   [1.2609 µs 1.3517 µs 1.5102 µs]
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by glorv** [comment]
@overvenus @Connor1996 PTAL, thanks~

**Review by Connor1996 — APPROVED** [review]
LGTM

File: components/tikv_util/src/memory.rs [hunk 1]
```diff
@@ -205,55 +216,31 @@ impl MemoryQuota {
...
-            match self.in_use.compare_exchange_weak(
-                in_use_bytes,
-                new_in_use_bytes,
-                Ordering::Relaxed,
-                Ordering::Relaxed,
-            ) {
-                Ok(_) => return,
-                Err(current) => in_use_bytes = current,
-            }
+        let bytes = bytes as isize;
```
Comment (overvenus) [inline components/tikv_util/src/memory.rs:238]: What if the bytes is larger than `isize::MAX`?

**Review by overvenus — CHANGES_REQUESTED** [review]
See https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=77b4b90699f5ba15ebe0e9c759f7aaba

File: components/tikv_util/src/memory.rs [hunk 2]
```diff
@@ -205,55 +216,34 @@ impl MemoryQuota {
     }
```
Comment (overvenus) [inline components/tikv_util/src/memory.rs:216]: Please check the numeric bound.

Comment (glorv) [reply components/tikv_util/src/memory.rs:221]: done

File: components/tikv_util/src/memory.rs [hunk 3]
```diff
@@ -200,60 +211,44 @@ impl MemoryQuota {
...
-                in_use_bytes,
-                new_in_use_bytes,
-                Ordering::Relaxed,
-                Ordering::Relaxed,
-            ) {
-                Ok(_) => return,
-                Err(current) => in_use_bytes = current,
-            }
-        }
+        self.in_use.fetch_add(bytes as isize, Ordering::Relaxed);
```
Comment (overvenus) [inline components/tikv_util/src/memory.rs:224]: Another issue

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=69bef8ec9096033e89d90b3c6b86d16a

**Review by overvenus — COMMENTED** [review]
It's better to check numeric bounds with every method call.

File: components/tikv_util/src/memory.rs (hunk 3 above)
Comment (glorv) [reply components/tikv_util/src/memory.rs:224]: But the previous implementation also has this issue that the value of `in_use` can overflow.

Comment (overvenus) [reply components/tikv_util/src/memory.rs:224]: The second alloc will return an error in the previous implementation.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5da24bccc3031d787b7541a841247663

---

It's okay that sometimes `in_use` is larger than `capacity`, but it's not okay that `alloc` continues returning success.

File: components/tikv_util/src/memory.rs (hunk 1 above)
Comment (glorv) [reply components/tikv_util/src/memory.rs:259]: Done. In order to handle overflow, I added a hard limit for a single alloc call.

**Comment by glorv** [comment]
@overvenus @Connor1996  PTAL again, thanks~

**Review by Connor1996 — APPROVED** [review]
LGTM

File: components/tikv_util/src/memory.rs [hunk 4]
```diff
@@ -211,23 +223,22 @@ impl MemoryQuota {
...
-            capacity = isize::MAX as usize;
-        }
+    pub fn set_capacity(&self, capacity: usize) {
+        let capacity = Self::adjust_capacity(capacity);
         self.capacity.store(capacity, Ordering::Relaxed);
     }
 
     pub fn alloc_force(&self, bytes: usize) {
+        if bytes > MAX_MEMORY_ALLOC_SIZE {
+            return;
```
Comment (overvenus) [inline components/tikv_util/src/memory.rs:233]: Can we add some log or metrics here?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: components/tikv_util/src/memory.rs (hunk 4 above)
Comment (glorv) [reply components/tikv_util/src/memory.rs]: Done. Add a warn log for oversized alloc.  Maybe we can also add a metrics for the alloc size distribution histogram later if we really need it.

**Review by overvenus — APPROVED** [review]
LGTM

**Pull request merged**